E2e k8bps (take #2) + Update to V1

jayunit100 opened the pull request.

jayunit100 commented:
Hi @quinton-hoole no rush, but just FYI, I reopened here.  I know you guys have some other stuff on the plate right now.   :)

jayunit100 commented:
note, this is tested locally - but last time the same feature broke GCE.  so may want to test it on GCE before merge if theres a free cluster i can help with that.

quinton-hoole commented:
Thanks Jay.  I'm not ignoring you, don't worry :-)

jayunit100 commented:
hi... of the commits in here is pretty critical k8bps update which the poor canonical guys had to do a hacky workaround to https://github.com/GoogleCloudPlatform/kubernetes/commit/f85e6bcf74710d7744691c357da065bd1f698287 .....  if this will be moved forward several days im happy to split it out or just modify the test so its `PIt` so it doesnt break things, and we can re-enable it after 1.0 release?

chuckbutler commented on examples/k8petstore/k8petstore.sh:
```diff
@@ -16,16 +16,32 @@
 
 echo "WRITING KUBE FILES , will overwrite the jsons, then testing pods. is kube clean ready to go?"
 
+
```
Missing a control header to keep the script from running away from the user. Typically this is done with a header like set -eu at the top of the file. Where there are cases you want errors to go without bailing on the user, set +e for the context of that command.

chuckbutler commented on examples/k8petstore/k8petstore.sh:
```diff
@@ -16,16 +16,32 @@
+
+echo "Note that, when running, If public IPs fail, you can use localhost:8080/api/v1beta1/proxy/services/frontend/llen?namespace=$NS and modify this script as necessary."
 
+kubectl="${1:-$kubectl}"
```
The params are a great start, but the fact they are positional makes it awkward to pass certain flags while leaving the rest at default.  However I'm not sure how to do an argparse style parser in bash.

chuckbutler commented on examples/k8petstore/k8petstore.sh:
```diff
@@ -188,53 +204,80 @@ cat << EOF > slave-rc.json
-	    echo "curl : $i"
-            curl "$PUBLIC_IP:3000/llen" >> result
+	for i in `seq 1 $TEST_SECONDS`;
+	 do
```
I'm wondering if we dont fire up something like SEIGE or HTTPerf to load test the service in paralell, vs doing X curl-calls against the petstore. This would give us slightly more meaningful results as to the resiliency of the service under load and would complete in a fraction of the time.

https://registry.hub.docker.com/u/jcostello84/siege-loadtest/ - seems like a good candidate for this, and the output can be parsed for reporting elsewhere.

chuckbutler commented:
@jayunit100  Thanks for reaching out. I took some time to deploy this and run through the revisions. I left comments inline on areas I identified that I feel we can improve and turn this basic test into a more meaningful structural test of the ecosystem/pods.

The only nit that I found was missing the set -eu at the top of the file allowing it to run away with itself / consume undeclared variables. 

Thanks for this however! the parameterized petstore is indeed a welcome addition for the action i built to consume it.

jayunit100 commented:
@chuckbutler okay, im going to create a separate PR for the bash improvements and maybe we can collaborate on those changes.  Then i will reincorporate that into this patch as need be.  great feedback!

quinton-hoole commented on examples/k8petstore/k8petstore.sh:
```diff
@@ -16,16 +16,32 @@
 FE="1"                # amount of Web server  
 LG="1"                # amount of load generators
 SLAVE="1"             # amount of redis slaves 
+TEST_SECONDS="1000"   # 0 = Dont run tests, if > 0,  run tests for n seconds.
```
Why so long?  Most e2e's run in under a minute.

jayunit100 replied:
Just a default.  Its not relevant to e2e at all.

quinton-hoole commented on test/e2e/soak_k8petstore.go:
```diff
@@ -0,0 +1,218 @@
+//once that is done, we can unset the Pending ginkgoflags (PIt).
+var ip = "123.456.789.100"
+
+// i.e. after 50 trials, expect 3000 transactions... minimum we settle for is 500.
```
What does this comment refer to?

jayunit100 commented:
pushed and rebased some cleanup fixed the unclear coment... , thanks squeezing a look  in these busy times ! 
- let me know if anything else  .  
- Also iirc you prefer the wait.poll() vs channels for polling,  if you want i can redo it that way although its mostly the same either way to me :).

jayunit100 commented:
and ps... DO NOT MERGE im testing now :)

jayunit100 commented:
ive tested this a couple of times.  Its working for me.  Any other thoughts ?   Given that folks are busy, and its generally clean and functional and useful, I'd like to suggest we get it in , and if any problems, I can easily just set the scale test to `PIt` rather than revert.

jayunit100 commented on test/e2e/soak_k8petstore.go:
```diff
@@ -0,0 +1,218 @@
+// readTransactions reads # of transactions from the k8petstore web server endpoint.
+// for more details see the source of the k8petstore web server.
+
+func readTransactions(c *client.Client, ns string) (error, int) {
```
@caesarxuchao one trick for running w/o public ip, is here...  iff you are running inside the kube network, you can use this strategy (curl localhost:8080/k8petstore/proxy/services/frontend/llen).

caesarxuchao replied:
Hi @jayunit100, yes, but as you have pointed out, this requires the user to run inside the kube network, so it won't work if I'm running k8spetstore.sh from my local workstation : (

jayunit100 replied:
yup thats true - it assumes that your in the cluster when running.

jayunit100 commented:
I wonder if its possibly to ship this w/ v1 ? the reason being that after V1 we will be updating k8petstore to use NodePorts, so the regression testing will be valuable to have in place.

quinton-hoole commented:
Yes, thanks for this Jay.  I'll try to review this in the next day or two.

jayunit100 commented:
There are now examples.go tests.  just want to explain why this is separate... 
This is a soak test so not sure if it belongs there or not, plus its quite involved w/ polling etc.  
Any thoughts on this logic?  I can merge to examples.go if necessary, but im against that as this test is meant to be more than just an example.

jayunit100 commented:
After thinking some more about it, I like this as a separate example, because it is a soak test - the examples are really just testing the examples/ for basic functionality (i.e. that the jsons compile and that the apps can do basic functions).

So, ready for final review imo !

timothysc commented:
@roberthbailey , @ixdy - could we get a reassign given that quinton is on vacation.

roberthbailey commented:
Re-assigned to @ixdy

timothysc commented:
Again, this is languishing.

jayunit100 commented:
looks like it needs a k8s.io rebase . will do this now

jayunit100 committed:
K8PetStore E2E test, rebased and cleaned (10/5, 10/6)

jayunit100 commented:
@kubernetes/goog-testing okay, this is ready to go now, its been rebased and retested.   hope we can get a final review and merge soon !

``` 
Aug 25 14:46:10.899: INFO: Now waiting 500 seconds to see progress (transactions  > 3)                                                                                                            [20/1848]
Aug 25 14:46:12.901: INFO: 2015-08-25 14:46:12.901931993 -0400 EDT == 3 total petstore transactions stored into redis. ==
Aug 25 14:46:14.901: INFO: 2015-08-25 14:46:14.901909125 -0400 EDT == 3 total petstore transactions stored into redis. ==
Aug 25 14:46:16.901: INFO: 2015-08-25 14:46:16.901768311 -0400 EDT == 3 total petstore transactions stored into redis. ==
Aug 25 14:46:18.902: INFO: 2015-08-25 14:46:18.902098878 -0400 EDT == 3 total petstore transactions stored into redis. ==
Aug 25 14:46:20.901: INFO: 2015-08-25 14:46:20.901891239 -0400 EDT == 168 total petstore transactions stored into redis. ==
Aug 25 14:46:22.901: INFO: 2015-08-25 14:46:22.901981933 -0400 EDT == 449 total petstore transactions stored into redis. ==

```

jayunit100 commented:
Can someone let me know why the "Jenkins GCE e2e" job failed?  It may or  may not be a flake.

wojtek-t commented:
The test that failed is:
Kubernetes e2e suite.Services should serve a basic endpoint from pods
This looks like a flake to me.

@k8s-bot test this please

jayunit100 commented:
thank @wojtek-t .  Is it ok for us to add K8PetStore to the E2E suite to make sure its running?  I realize  there are a lot of filters on there.

timothysc commented:
@kubernetes/goog-testing - could we get a reassign on this one and knock it off our backlog.  It's been reviewed multiple times now and dates back to May.

quinton-hoole commented:
Sure, I can take it.  @ixdy, FYI.

jayunit100 commented:
Thanks for looking @quinton-hoole ...  can you check in the E2E CI and confirm that the test runs? It should print something like this: 

```
Aug 25 14:46:10.899: INFO: ... Done starting k8petstore
Aug 25 14:46:10.899: INFO: Start polling, timeout is 300 seconds
Aug 25 14:46:10.899: INFO: Now waiting 500 seconds to see progress (transactions  > 3)
Aug 25 14:46:12.901: INFO: 2015-08-25 14:46:12.901931993 -0400 EDT == 3 total petstore transactions stored into redis. ==
Aug 25 14:46:14.901: INFO: 2015-08-25 14:46:14.901909125 -0400 EDT == 3 total petstore transactions stored into redis. ==
Aug 25 14:46:16.901: INFO: 2015-08-25 14:46:16.901768311 -0400 EDT == 3 total petstore transactions stored into redis. ==
Aug 25 14:46:18.902: INFO: 2015-08-25 14:46:18.902098878 -0400 EDT == 3 total petstore transactions stored into redis. ==
Aug 25 14:46:20.901: INFO: 2015-08-25 14:46:20.901891239 -0400 EDT == 168 total petstore transactions stored into redis. ==
Aug 25 14:46:22.901: INFO: 2015-08-25 14:46:22.901981933 -0400 EDT == 449 total petstore transactions stored into redis. ==
```

That will confirm that its working properly and that its running as part of the regression.

quinton-hoole commented on test/e2e/soak_k8petstore.go:
```diff
@@ -0,0 +1,231 @@
+/*
```
Is there any reason why this should not go in test/e2e/examples.go?
Also, prefixing the file name with "soak_" is inconsistent with all other e2e tests.

jayunit100 replied:
The reason this is using an example as a mechanism to run a soak test.   Its not really a typcial examples.go test....  i can :
- rename the test file itself to something like `k8petstore.go` ? 
- or can move it if you'd move it into `examples.go`.  

I'd sort of lean to keep it separate but I realize this is a bit of a toss-up...

quinton-hoole replied:
OK, please rename to k8petstore.go, thanks.

quinton-hoole commented on test/e2e/soak_k8petstore.go:
```diff
@@ -0,0 +1,231 @@
+	Ω(totalTransactions).Should(BeNumerically(">", minExpected))
+}
+
+var _ = Describe("K8PetStore (succesfull launching and transactions in the k8petstore application)", func() {
```
Just "Pet Store" please.

quinton-hoole commented on test/e2e/soak_k8petstore.go:
```diff
@@ -0,0 +1,231 @@
+
+	//This must run on a machine which is in the kube-proxy ring.  Otherwise, the publicIP binding will fail.
+	//The IP Below ~ letter->number cipher for k8petstore (most likely it won't be bound by any other process)
+	var ip = "165.201.92.15"
```
This won't work in general.   What does "letter->number cipher" mean?

jayunit100 replied:
This IP isn't needed for NodePorts, but it is a parameter we keep in the test as a parameter which can be used by the testing function.  Since the only running test function doesn't use the IP at all, I can delete the cipher entirely for now I guess to simplify things.

quinton-hoole commented on test/e2e/soak_k8petstore.go:
```diff
@@ -0,0 +1,231 @@
+
+	//On a single node cluster, we expect about 10 transactions every second on average.
+	//admittedly, this is a very rough estimate given that the ETL is done in batches...
+	It(fmt.Sprintf("Should be FUNCTIONALLY able to successfully acquire 500 transactions per minion"), func() {
```
Why sprintf?  It's superfluous here.

quinton-hoole replied:
Can we please make the number of transactions, and timeout configurable?  And please reword this to be more succint e.g. 
"should acquire at least %v transactions in %v seconds"

quinton-hoole commented on test/e2e/soak_k8petstore.go:
```diff
@@ -0,0 +1,231 @@
+	})
+
+	//nThis test takes a few minutes... for simple CI setups testing pure functionality, filter it out.
+	PIt(fmt.Sprintf("should SCALE to support generating and storing several thousands of transactions per node"), func() {
```
Why sprintf? It's superfluous here.

quinton-hoole replied:
Could we be more succinct and specific here, e.g:
"should scale to %v transactions per node"

jayunit100 replied:
long term, that exact number may vary.  but in any case, ill update it to be precise for now.  
over time id like this to be user configurable and based on some cluster heuristics / metadata if that is possible.

quinton-hoole commented on test/e2e/soak_k8petstore.go:
```diff
@@ -0,0 +1,231 @@
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Building a namespace api object")
+		namespace, err = createTestingNS("k8petstore-soak", c)
```
Please use Framework for this:
https://github.com/kubernetes/kubernetes/blob/95a60d60856987f36fc666252d1e74ca9c5d377e/test/e2e/framework.go#L45

quinton-hoole commented on test/e2e/soak_k8petstore.go:
```diff
@@ -0,0 +1,231 @@
+// transactions are acquired in a maximum of "max_seconds".
+func runK8petstore(ip string, restServers int, loadGenerators int, c *unversioned.Client, ns string, minExpected int, maxSeconds int64) {
+
+	var root0 = absOrDie(filepath.Clean(filepath.Join(path.Base(os.Args[0]), "..")))
```
What is this for?  Can it be removed?

jayunit100 replied:
this is the path to the examples/ directory.   Renaming the var now so that its clear.

quinton-hoole commented on test/e2e/soak_k8petstore.go:
```diff
@@ -0,0 +1,231 @@
+	cmd := exec.Command(
+		k8bps,
+		"kubectl", //for dev, replace w/ "cluster/kubectl.sh"
+		"r.2.8.19",
```
What is this?  Please make it a named constant, with comments.

quinton-hoole commented on test/e2e/soak_k8petstore.go:
```diff
@@ -0,0 +1,231 @@
+		k8bps,
+		"kubectl", //for dev, replace w/ "cluster/kubectl.sh"
+		"r.2.8.19",
+		"1000",
```
What it this? Please make this a named constant.

quinton-hoole commented on test/e2e/soak_k8petstore.go:
```diff
@@ -0,0 +1,231 @@
+
+// runK8petstore runs the k8petstore application, bound to external ip "ip", and polls it to assert that "min_expected"
+// transactions are acquired in a maximum of "max_seconds".
+func runK8petstore(ip string, restServers int, loadGenerators int, c *unversioned.Client, ns string, minExpected int, maxSeconds int64) {
```
Why do you need an int64 rather than an int here?  A plain int can hold more than 80 years worth of seconds :-)

quinton-hoole commented on test/e2e/soak_k8petstore.go:
```diff
@@ -0,0 +1,231 @@
+
+	var root0 = absOrDie(filepath.Clean(filepath.Join(path.Base(os.Args[0]), "..")))
+	var err error = nil
+	k8bps := filepath.Join(root0, "examples/k8petstore/k8petstore-nodeport.sh")
```
Please come up with a better variable name here.  BPS traditionally means something like bits per second.  What does it imply here?  Bound Pet Store?

jayunit100 replied:
fixing (see above)

quinton-hoole commented on test/e2e/soak_k8petstore.go:
```diff
@@ -0,0 +1,231 @@
+		fmt.Printf("load generators / rest generators [ %v  /  %v ] ", loadGenerators, restServers)
+
+		//At least 500 transactions should be acquired within a short time period.
+		runK8petstore(ip, restServers, loadGenerators, c, ns, 500, 60*5)
```
To be run routinely, this test will need to complete in about one minute.  Can we reduce the duration to 1 minute please?  Otherwise it will need to be pulled out into a separate slow-running set of tests, and won't be able to be included in the parallel e2e tests that are run against every PR pre-merge.

jayunit100 replied:
also addressed by moving transaction limit down to 50/60 seconds.

quinton-hoole commented on test/e2e/soak_k8petstore.go:
```diff
@@ -0,0 +1,231 @@
+		//the more generators -> the more transactions and the more bursty CPU used per minion.
+		var loadGenerators int = minionCount * 2
+		var restServers int = minionCount
+		//var min_expected int = trials * 10 * minionCount // more minions --> higher expected # of transactions.
```
Please remove debug comments.

quinton-hoole commented:
Unfortunately this test is for some reason being excluded from our automated per-PR tests, and I don't fully understand why.  The answer is burried here:

https://github.com/kubernetes/kubernetes/blob/master/hack/jenkins/e2e.sh#L299

Presumably the current name of your test ("K8PetStore (succesfull launching and transactions in the k8petstore application) should SCALE to support generating and storing several thousands of transactions per node") is accidentally matching one of the following regular expressions:

https://github.com/kubernetes/kubernetes/blob/master/hack/jenkins/e2e.sh#L90
https://github.com/kubernetes/kubernetes/blob/master/hack/jenkins/e2e.sh#L90
https://github.com/kubernetes/kubernetes/blob/master/hack/jenkins/e2e.sh#L124

... but I can't immediately see which one.

But you're probably going to rename the test anyway (this is one example of why it's good to be succinct :-) . So lets check again once you've addressed the code review comments.

jayunit100 commented:
Thanks for the review quinton.  

I finished addressing all comments and pushed the changes, mainly

- I removed any reference to an IP.  This test doesnt need the chiper IP thing, even though related tests we'd like to would benefit from testing explicit IPs, its not needed in here.
- int64, and a couple of stray comment lines removed, test names cleaned up.
- reduced test load to 50 transactions, 60 seconds.  
- replaced all explicit namespace creation w/ the framework API.  This is a new development since the original PR.
- also replaced the `root0` file path.  the new API supports use of kube `RepoRoot` option, wihch has the equivalent behaviour. 

One thing: I didn't move the test into  `examples.go`, bc its meant to be a full soak test of kubernetes,  not really a examples verification test. If rename is required, just LMK i'll gladly move the file and re-commit, but IMO it will live a  life of its own and shouldnt be coupled to examples/ necessarily.   Long term maybe we can even implemment all of k8petstore in Golang...

quinton-hoole commented on test/e2e/soak_k8petstore.go:
```diff
@@ -0,0 +1,181 @@
+	Ω(totalTransactions).Should(BeNumerically(">", minExpected))
+}
+
+var _ = Describe("K8 Pet Store", func() {
```
As per previous review, just "Pet Store" please. All tests are for Kubernetes, so the prefix is unnecessary.  As an aside, the abbreviation for "Kubernetes" is "k8s".

jayunit100 commented:
Okay, done -  renamed to `test/e2e/k8petstore.go`.

quinton-hoole commented on test/e2e/soak_k8petstore.go:
```diff
@@ -0,0 +1,181 @@
+
+	//Lets poll until we reach min_expected transactions
+	totalTransactions := 0
+	startupTimeoutSeconds := 500
```
This is almost 10 minutes, and we want the test to complete in 1 minute.  Can we reduce this?

jayunit100 replied:
Yes, I can reduce it to a 60 second time for the cold start.

quinton-hoole commented on test/e2e/soak_k8petstore.go:
```diff
@@ -0,0 +1,181 @@
+		var restServers int = minionCount
+		fmt.Printf("load generators / rest generators [ %v  /  %v ] ", loadGenerators, restServers)
+		//At least 500 transactions should be acquired within a short time period.
+		runK8petstore(restServers, loadGenerators, f.Client, f.Namespace.Name, 50, 60)
```
Please make the 50 and 60 named configurable constants, as per previous review.  You'll need to fix the above "It" statement too, which contains duplicates of those numbers.

jayunit100 replied:
Earlier, I thought I responded but maybe I didn't....   In any case, yes i can make these constants if necessary, I was a little opposed to it b/c the test itself is supposed to define these.  But for now I guess its ok, ill implement and re-test , then ill  push ;

quinton-hoole commented on test/e2e/k8petstore.go:
```diff
@@ -0,0 +1,179 @@
+
+	//On a single node cluster, we expect about 10 transactions every second on average.
+	//admittedly, this is a very rough estimate given that the ETL is done in batches...
+	It("Should  scale to 50 transactions in one minute", func() {
```
"It" clauses start with lower case please, as they are concatenated with the "Describe clause to form a valid sentence, which is used in the test output and reports.  Also, fix the double-space after "should".

jayunit100 commented:
ok fixed `It` should doublespace, rebased and pushed;  :)

jayunit100 commented:
Okay ... 
- made the times into constants at the top.  not sure its the best idea for later tests, but for this one, i guess its ok.  later when i add the scaling component ill move to hashmaps.
- updated the associated log statements.
- trimmed the cold start timeout to 30 seconds, in general it should run faster than that.

jayunit100 commented:
I've also added one new thing - it uses the `kubectl-path` option, rather than the hardcoded "kubectl" string.  this is a new option which was added after i originally wrote this test.

quinton-hoole commented on test/e2e/k8petstore.go:
```diff
@@ -0,0 +1,182 @@
+		var loadGenerators int = minionCount
+		var restServers int = minionCount
+		fmt.Printf("load generators / rest generators [ %v  /  %v ] ", loadGenerators, restServers)
+		//At least 500 transactions should be acquired within a short time period.
```
This comment is no longer correct, given the current value of k8bpsSmokeTestTransactions.

jayunit100 replied:
500 transactions should be acquired rapidly.  That fact doesn't change with this parameters.  The parameters themselves should based on the fact that, the batches are sent in several hundred at a time, so as soon as the first data generator runs, we shoudl have several 100s of transactions.

quinton-hoole commented on test/e2e/k8petstore.go:
```diff
@@ -0,0 +1,182 @@
+	var minionCount int
+	f := NewFramework("petstore")
+
+	//On a single node cluster, we expect about 10 transactions every second on average.
```
This comment is inconsistent with the test, which expects 50 transactions in 60 seconds?

jayunit100 replied:
why is it inconsistent ?  50 should happen within about 5 seconds or so, once the containers are pulled down and fired up....

jayunit100 commented:
It appears the two new comments are noting that the comments are referring to "upper bounds", where as the constants are "lower bounds" .  That is ok, IMO, b/c in a smoke test parameterization ( `k8bpsSmokeTestTransactions`) we are just verifying that the system is able to support a minimal transaction load.  The secondary tests will reallly push it to the limit.  This is just the first step of getting the soak test into the continuous-integration pipeline.

quinton-hoole commented:
You should clarify the comments in the code, as they are definitely confusing to others (including me) as they currently stand.

jayunit100 commented:
Ok I can do that; 
Maybe I can even remove them or minimize them where possible.

jayunit100 commented:
Updated the comments.   Sorry this took so long, I actually thought about it for  a while and just deleted one of the comments, and added some more info to the test declaration.  I think that should be sufficient.

timothysc commented:
/cc @danmcp automated saturation testing.

jayunit100 commented:
@quinton-hoole , pending rebase, does this otherwise look okay?

quinton-hoole commented on examples/k8petstore/k8petstore-nodeport.sh:
```diff
@@ -245,20 +246,21 @@ function getIP {
     echo -e "List of nodes' IP addresses:\n$NODES_IP"
     echo -e "Node IP $TEST_IP will be used in the tests in this script"
     printf '%*s\n\n\n\n' "${COLUMNS:-$(tput cols)}" '' | tr ' ' =
-  fi
```
This breaks indentation.

quinton-hoole commented on examples/k8petstore/k8petstore-nodeport.sh:
```diff
@@ -245,20 +246,21 @@ function getIP {
 }
 
 function getNodePort {
-NODE_PORT=$($kubectl get services/frontend -t='{{(index .spec.ports 0).nodePort}}')
```
Indentation broken in this function.

quinton-hoole commented on test/e2e/k8petstore.go:
```diff
@@ -0,0 +1,179 @@
+	//How long until the FIRST transaction...
+	startupTimeout := time.After(time.Duration(k8bpsStartupTimeout))
+
+	//How long we will wait until we reach the nth (ie. 500th) transaction
```
You're still incorrectly mentioning 500 transaction here.  Please self-review your code before submitting for others to review.

jayunit100 replied:
Should I just delete the comment?  Its just meant to clarify what is meant by "nth", probably not necessary.  I don't think there is a true "correct" number to put here....

quinton-hoole commented on test/e2e/k8petstore.go:
```diff
@@ -0,0 +1,179 @@
+	for {
+		select {
+		case <-transactionsCompleteTimeout:
+			//i.e. if we dont have n*1000 transactions, fail the test.
```
What is 'n', and why times 1000?

jayunit100 replied:
In general, its expected that this test will confirm there are several 1000 transactions.  Should I just delete the comment?

jayunit100 commented:
Ok.  I've cleaned up the comments some more, and re-reviewed  this myself, changes pushed...

jayunit100 commented:
@timothysc can you do one (hopefully last) a final review of this, then i'll re-ping  again.

jayunit100 commented:
@k8s-bot test this please.

timothysc commented:
ping @wojtek-t I can't apply labels, could you take?

wojtek-t commented on test/e2e/k8petstore.go:
```diff
@@ -0,0 +1,171 @@
+
+// readTransactions reads # of transactions from the k8petstore web server endpoint.
+// for more details see the source of the k8petstore web server.
+
```
please remove this empty line

wojtek-t commented on test/e2e/k8petstore.go:
```diff
@@ -0,0 +1,171 @@
+	k8bpsDontRunTest      = "0"              // Don't bother embedded test.
+	k8bpsStartupTimeout   = 30 * time.Second // Amount of elapsed time before petstore transactions are being stored.
+
+	//Constants for the first test. We can make this a hashmap once we add scale tests to it.
```
please add space before "Constants" (i.e. "// Constants ...") - for consistency

wojtek-t commented on test/e2e/k8petstore.go:
```diff
@@ -0,0 +1,171 @@
+	cmd.Stderr = os.Stderr
+
+	Logf("Starting k8petstore application....")
+	//Run the k8petstore app, and log / fail if it returns any errors.
```
please add space after "//"

same in other places

wojtek-t commented:
@jayunit100 - please squash the commits

jayunit100 commented:
@wojtek-t ok, cleaned the newlines/comments + rebased.  Thanks for the review...!

wojtek-t commented on test/e2e/k8petstore.go:
```diff
@@ -0,0 +1,170 @@
+	if err = cmd.Start(); err != nil {
+		log.Fatal(err)
+	}
+	//Make sure there are no command errors.
```
Can you please add space after "//" in all places below (for consistency)?

wojtek-t commented on test/e2e/k8petstore.go:
```diff
@@ -0,0 +1,170 @@
+	var minionCount int
+	f := NewFramework("petstore")
+
+	It(fmt.Sprintf("should scale to persist a nominal number ( %v ) of transactions in %v seconds", k8bpsSmokeTestTransactions, k8bpsSmokeTestTimeout), func() {
```
Can you please add "[Skipped][Example]" to the name of the test (similarly to what we have e.g. here:
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/examples.go#L250

wojtek-t commented on test/e2e/k8petstore.go:
```diff
@@ -0,0 +1,170 @@
+		Expect(err).NotTo(HaveOccurred())
+		minionCount = len(minions.Items)
+
+		var loadGenerators int = minionCount
```
why not just:
"loadGenerators := minionCount"
?

(same for restServers)

jayunit100 commented:
@wojtek-t ok, implemented the `=` -> `:=` change , and fixed `// `+spaces. pushed and squashed.

wojtek-t commented:
LGTM
